File: app/src/main/java/com/qust/lesson/LessonTableViewModel.java
package com.qust.lesson;

import android.app.Application;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.MutableLiveData;

import com.qust.assistant.R;
import com.qust.assistant.util.LogUtil;
import com.qust.assistant.util.SettingUtil;
import com.qust.utils.FileUtils;

import java.io.File;
import java.io.IOException;
import java.text.ParseException;
import java.util.Calendar;
import java.util.Date;
import java.util.TimeZone;

public class LessonTableViewModel extends AndroidViewModel{
	
	private static volatile LessonTableViewModel INSTANCE;
	
	
	/**
	 * 课程时间表
	 */
	public static final String[][][] LESSON_TIME_TEXT = new String[][][]{
		{
			{"08:00", "09:00", "10:10", "11:10", "13:30", "14:30", "15:40", "16:40", "18:00", "19:00"},
			{"08:50", "09:50", "11:00", "12:00", "14:20", "15:20", "16:30", "17:30", "18:50", "19:50"}
		},{
			{"08:00", "09:00", "10:10", "11:10", "14:00", "15:00", "16:10", "17:10", "18:30", "19:30"},
			{"08:50", "09:50", "11:00", "12:00", "14:50", "15:50", "17:00", "18:00", "19:20", "20:20"}
		}
	};
	
	/**
	 * 课程时间差 (单位：分钟)
	 */
	public static final int[][] LESSON_TIME = {
		// 冬季
		{ 0, 60, 70, 60, 140, 60, 70, 60, 80, 60 },
		// 夏季
		{ 0, 60, 70, 60, 170, 60, 70, 60, 80, 60 }
	};
	
	
	/**
	 * 课程时间差
	 */
	private int[] lessonTime;
	
	/**
	 * 上下课时间文本
	 */
	private String[][] lessonTimeText;
	
	/**
	 * 当前时间表
	 */
	private int currentTime;
	
	/**
	 * 当前周 (从1开始)
	 */
	private int currentWeek;
	
	/**
	 * 当前星期 ( 0-6, 周一 —— 周日)
	 */
	private int dayOfWeek;
	
	private LessonTable lessonTable;
	
	/**
	 * 是否需要更新课表UI（课表信息变化时的回调）
	 */
	private final MutableLiveData<LessonTable> lessonTableLiveData;
	
	
	public static LessonTableViewModel getInstance(Context context){
		if(INSTANCE == null){
			synchronized(LessonTableViewModel.class){
				if(INSTANCE == null){
					INSTANCE = new LessonTableViewModel((Application)context.getApplicationContext());
				}
			}
		}
		return INSTANCE;
	}
	
	
	public LessonTableViewModel(@NonNull Application application){
		super(application);
		
		loadLesson(application);
		
		updateDate();
		
		currentTime = SettingUtil.getInt(application.getString(R.string.KEY_TIME_TABLE), 0);
		lessonTime = LESSON_TIME[currentTime];
		lessonTimeText = LESSON_TIME_TEXT[currentTime];
		lessonTableLiveData = new MutableLiveData<>(lessonTable);
	}

	
	public MutableLiveData<LessonTable> getLessonTableLiveData(){ return lessonTableLiveData; }
	
	
	/**
	 * 更新日期信息
	 */
	public void updateDate(){
		Calendar c = Calendar.getInstance(TimeZone.getTimeZone("GMT+8:00"));
		c.setFirstDayOfWeek(Calendar.MONDAY);
		
		Calendar c2 = (Calendar)c.clone();
		
		c.setTime(lessonTable.getStartDay());
		int startWeek = c.get(Calendar.WEEK_OF_YEAR);
		
		int day = c2.get(Calendar.DAY_OF_WEEK);
		dayOfWeek = (day == Calendar.SUNDAY ? 6 : day - 2);
		currentWeek = Math.max(1, c2.get(Calendar.WEEK_OF_YEAR) - startWeek + 1);
	}
	
	
	/**
	 * 获取当前周（从1开始）
	 */
	public int getCurrentWeek(){
		return currentWeek;
	}
	
	/**
	 * 获取当前是星期几(0-6, 周一 —— 周日)
	 */
	public int getDayOfWeek(){
		return dayOfWeek;
	}
	
	
	/**
	 * 设置开学时间
	 */
	public synchronized void setStartDay(@NonNull String _startDay) throws ParseException{
		lessonTable.setStartDay(_startDay);
		updateDate();
		lessonTableLiveData.postValue(lessonTable);
		saveLesson();
	}
	
	/**
	 * 设置开学时间
	 */
	public synchronized void setStartDay(@NonNull Date _startDay){
		lessonTable.setStartDay(_startDay);
		updateDate();
		lessonTableLiveData.postValue(lessonTable);
		saveLesson();
	}
	
	/**
	 * 设置学期总周数
	 */
	public synchronized void setTotalWeek(int _totalWeek){
		lessonTable.setTotalWeek(_totalWeek);
		lessonTableLiveData.postValue(lessonTable);
		saveLesson();
	}
	
	/**
	 * 设置课表
	 */
	public synchronized void setLessonGroups(LessonGroup[][] _lessonGroups){
		lessonTable.setLessons(_lessonGroups);
		lessonTableLiveData.postValue(lessonTable);
		saveLesson();
	}
	
	/**
	 * 设置课程时间表
	 */
	public void setCurrentTime(int index){
		currentTime = index;
		lessonTime = LESSON_TIME[index];
		lessonTimeText = LESSON_TIME_TEXT[index];
		lessonTableLiveData.postValue(lessonTable);
		SettingUtil.put(getApplication().getString(R.string.KEY_TIME_TABLE), index);
	}
	
	
	/**
	 * 获取开学时间
	 */
	public Date getStartDay(){
		return lessonTable.getStartDay();
	}
	
	/**
	 * 获取学期总周数
	 */
	public int getTotalWeek(){
		return lessonTable.getTotalWeek();
	}
	
	/**
	 * 获取课表
	 * @return
	 */
	public LessonGroup[][] getLessonGroups(){
		return lessonTable.getLessons();
	}
	
	/**
	 * 获取全部课表信息
	 * @return
	 */
	public LessonTable getLessonTable(){ return lessonTable; }
	
	/**
	 * 获取当前时间表
	 * @return
	 */
	public int getCurrentTime(){
		return currentTime;
	}
	
	/**
	 * 获取课程时间差
	 */
	public int[] getLessonTime(){
		return lessonTime;
	}
	
	/**
	 * 获取课程时间文本
	 */
	public String[][] getLessonTimeText(){
		return lessonTimeText;
	}
	
	
	/**
	 * 设置并保存课程表信息
	 */
	public void saveLessonData(LessonTable _lessonTable){
		lessonTable = _lessonTable;
		lessonTableLiveData.postValue(lessonTable);
		saveLesson();
	}
	
	
	/**
	 * 从本地文件初始化课表
	 */
	private void loadLesson(@NonNull Context context){
		File dataFile = new File(context.getFilesDir(),"lessonTables");
		
		if(dataFile.exists()){
			File[] lessonTables = dataFile.listFiles();
			if(lessonTables != null){
				for(File lessonTableFile: lessonTables){
					try{
						lessonTable = (LessonTable) FileUtils.loadData(lessonTableFile);
						break;
					}catch(IOException | ClassNotFoundException e){
						LogUtil.Log(e);
						break;
					}
				}
			}
		}else{
			dataFile.mkdirs();
		}
		
		if(lessonTable == null){
			lessonTable = new LessonTable();
		}
	}
	
	/**
	 * 序列化储存课表数据
	 */
	private void saveLesson(){
		
		// 去除上课周数为0的课程
		LessonGroup[][] lessonGroups = lessonTable.getLessons();
		for(LessonGroup[] lessonGroup : lessonGroups){
			for(int timeSlot = 0; timeSlot < lessonGroup.length; timeSlot++){
				LessonGroup group = lessonGroup[timeSlot];
				if(group == null) continue;
				if(group.lessons.length == 0) lessonGroup[timeSlot] = null;
				for(int k = 0; k < group.lessons.length; k++){
					if(group.lessons[k].week == 0){
						group.removeLesson(k);
						break;
					}
				}
			}
		}
		
		File dataFile = new File(getApplication().getFilesDir(),"lessonTables");
		if(!dataFile.exists() && !dataFile.mkdirs()) return;
		try{
			FileUtils.saveData(new File(dataFile, "lessonTable"), lessonTable);
		}catch(IOException e){
			LogUtil.Log(e);
		}
	}
}

File: app/src/main/java/com/qust/helper/viewmodel/LessonTableViewModel.kt
package com.qust.helper.viewmodel

import android.app.Application
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.AndroidViewModel
import com.qust.helper.data.Keys
import com.qust.helper.data.Setting
import com.qust.helper.data.lesson.LessonGroup
import com.qust.helper.data.lesson.LessonTable
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.decodeFromStream
import kotlinx.serialization.json.encodeToStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.util.Calendar

class LessonTableViewModel(application: Application): AndroidViewModel(application) {

	/**
	 * 当前时间表
	 */
	var currentTimeTable = mutableIntStateOf(Setting.getInt(Keys.KEY_TIME_TABLE, 0))

	/**
	 * 当前周 (从1开始)
	 */
	var currentWeek = mutableIntStateOf(0)

	/**
	 * 当前星期 ( 0-6, 周一 —— 周日)
	 */
	var dayOfWeek = mutableIntStateOf(0)

	/**
	 * 显示所有课程
	 */
	var showAllLesson = mutableStateOf(Setting.getBoolean(Keys.KEY_SHOW_ALL_LESSON, true))

	/**
	 * 隐藏已结课课程
	 */
	var hideFinishLesson =  mutableStateOf(Setting.getBoolean(Keys.KEY_HIDE_FINISH_LESSON, false))

	/**
	 * 隐藏教师
	 */
	var hideTeacher = mutableStateOf(Setting.getBoolean(Keys.KEY_HIDE_TEACHER, true))


	var lessonTable = mutableStateOf(LessonTable(
		totalWeek = 20,
		lessons = Array(7){ arrayOfNulls(10) }
	))

	private var dataFile: File

	init {
		dataFile = File(application.filesDir, "lessonTables")
		loadLesson()
		updateDate()
//		lessonTime = LessonTableViewModel.LESSON_TIME.get(currentTimeTable)
//		lessonTimeText = LessonTableViewModel.LESSON_TIME_TEXT.get(currentTimeTable)
	}

	/**
	 * 更新日期信息
	 */
	fun updateDate() {
		val currentDay: Calendar = Calendar.getInstance().also { it.firstDayOfWeek = Calendar.MONDAY }
		val startDay: Calendar = (currentDay.clone() as Calendar).also { it.time = lessonTable.value.startDay }

		dayOfWeek.intValue = currentDay[Calendar.DAY_OF_WEEK].let { if(it == Calendar.SUNDAY) 6 else it - 2 }
		currentWeek.intValue = (currentDay[Calendar.WEEK_OF_YEAR] - startDay[Calendar.WEEK_OF_YEAR] + 1).coerceAtLeast(1)
	}

	/**
	 * 从本地文件初始化课表
	 */
	@OptIn(ExperimentalSerializationApi::class)
	private fun loadLesson() {
		if(dataFile.exists()) {
			dataFile.listFiles()?.let{
				for(file in it){
					if(file?.exists() == true) {
						try{
							FileInputStream(file).use { stream ->
								lessonTable.value = Json.decodeFromStream<LessonTable>(stream)
							}
							break
						}catch(_: Exception) { }
					}
				}
			}
		} else {
			dataFile.mkdirs()
		}
	}

	@OptIn(ExperimentalSerializationApi::class)
	fun saveLessonTable(): Boolean{

		// 去除上课周数为0的课程
		val lessonGroups: Array<Array<LessonGroup?>> = lessonTable.value.lessons
		for(dailyLesson in lessonGroups) {
			for(timeSlot in dailyLesson.indices) {
				val group: LessonGroup = dailyLesson[timeSlot] ?: continue
				if(group.lessons.isEmpty()){
					dailyLesson[timeSlot] = null
					continue
				}
				group.lessons = group.lessons.filter { it.week != 0L }.toTypedArray()
			}
		}

		if(!dataFile.exists()) dataFile.mkdirs()
		val file = File(dataFile, "lessonTable")
		return try{
			FileOutputStream(file).use { stream ->
				Json.encodeToStream<LessonTable>(lessonTable.value, stream)
			}
			true
		}catch(_: Exception) {
			false
		}
	}

}